### ISSUE TITLE
`highlight` with `extent` interferes with syntax highlighting

### ISSUE THREAD

**Issue by deffi** [typst/typst#2939]
### Description

Code:

```typst
#highlight(extent: 1mm)[```py for lorem in ipsum```]
```

Result:

![Result](https://github.com/typst/typst/assets/470243/40623144-12d5-49a4-a891-1246aa8527b1)

Probably related:
```typst
Lorem #highlight(extent:5mm)[ipsum] dolor
```

Result:

![Result](https://github.com/typst/typst/assets/470243/c147dca1-642b-4720-8807-a978b3e9afec)

The highlight is printed over the preceding text, but below the succeeding text.

Typst 0.10.0

### Operating system

Windows

### Typst version

- [X] I am using the latest version of Typst

**Comment by joserlopes** [comment]
Hello! Is it possible for me to look into this bug? Thank you!

**Comment by laurmaedje** [comment]
@joserlopes Yeah, sure. The relevant code is in `crates/typst/src/text/deco.rs`, the problem being that decorations are pushed one-by-one with the text items so that the decoration of the second text item is pushed later than the first text item. This would need to be changed somehow.

**Comment by Andrew15-5** [comment]
I thought to mention this in https://github.com/typst/typst/issues/5324, but practically it's about `extent`, so here is a similar example, but I need to have some text linked. `link` (as is [`underline`](https://github.com/typst/typst/issues/5324#issuecomment-2453377458) and other?) function creates normally invisible "parts," which are very apparent when (semi-transparent) `highlight` is applied:

```typ
#set highlight(extent: 0.1em, fill: green.transparentize(60%))
#highlight[Some #link(".")[linked] text.]
```

<img width="231" height="43" alt="Image" src="https://github.com/user-attachments/assets/12713118-df73-4a21-8ce4-2ce07d217399" />

By the way, @joserlopes, did you take a look at this already, any news?

**Comment by joserlopes** [comment]
No, I still haven't looked at this in any way, don't know if I ever will...
